javalin/src/main/java/io/javalin/core/validation/Validator.kt

```diff
@@ -6,107 +6,44 @@
 
 package io.javalin.core.validation
 
+import io.javalin.core.util.JavalinLogger
 import io.javalin.http.BadRequestResponse
+import io.javalin.http.InternalServerErrorResponse
 
-open class Validator<T>(val value: T?, val messagePrefix: String = "Value", val key: String = "Parameter") {
+enum class ValidationError { NULLCHECK_FAILED, TYPE_CONVERSION_FAILED, CUSTOM_CHECK_FAILED, DESERIALIZATION_FAILED }
 
-    data class Rule<T>(val fieldName: String, val test: (T) -> Boolean, val invalidMessage: String)
-
-    protected open var errors: MutableMap<String, MutableList<String>>? = null
-
-    protected val rules = mutableSetOf<Rule<T>>()
-
-    @JvmOverloads
-    open fun check(predicate: (T) -> Boolean, errorMessage: String = "Failed check"): Validator<T> {
-        rules.add(Rule(key, predicate, errorMessage))
-        return this
-    }
-
-    //These two options will fail fast but only provide the first failure.
-    fun get(): T = getOrNull() ?: throw BadRequestResponse("$messagePrefix cannot be null or empty")
+/**
+ * The non-nullable [Validator] uses [Rule] rules, but checks if value is null before calling them.
+ * The [check] method wraps its non-nullable predicate in a nullable predicate
+ */
+open class Validator<T>(value: T?, val fieldName: String) : BaseValidator<T>(value) {
 
-    fun getOrNull(): T? {
-        if (value == null) return null
-        return rules.find { !it.test.invoke(value) }?.let { throw BadRequestResponse("$messagePrefix invalid - ${it.invalidMessage}") } ?: value
+    init {
+        addRule(fieldName, { it != null }, {
+            JavalinLogger.info("Parameter '${fieldName}' cannot be null")
+            ValidationError.NULLCHECK_FAILED.name
+        })
     }
 
-    open fun isValid(): Boolean {
-        if (errors == null) {
-            validate()
-        }
-        return errors!!.isEmpty()
-    }
+    fun allowNullable() = NullableValidator(value, fieldName)
 
-    open fun hasError(): Boolean {
-        if (errors == null) {
-            validate()
-        }
-        return errors!!.isNotEmpty()
-    }
+    fun check(check: Check<T>, errorProvider: ErrorProvider) =
+        addRule(fieldName, { check(it!!) }, errorProvider) as Validator<T>
```

Comment:
Using it!! will just cause a null pointer exception here which will prevent the wellish formatted BadRequestResponse from being thrown as would likely be expected.

Reply:
In order to safely be able to pass these rules to a NullableValidator if you wanted to, you could create a wrapper.
```kotlin
fun <T?> required(predicate: (T) -> Boolean): NullableRule<T?> {
    return { 
        if (it == null) true else predicate(it)
     }
}
```

Reply:
> Using it!! will just cause a null pointer exception here which will prevent the wellish formatted BadRequestResponse from being thrown as would likely be expected.

The `init{}` block adds a rule which checks for this: 
```addRule(fieldName, { it != null }, RuleViolation.NULLCHECK_FAILED.name)```, 

So this test passes: 

```kotlin
app.get("/") { ctx -> ctx.queryParam<String>("my-qp").get() }
assertThat(http.get("/").body).isEqualTo("NULLCHECK_FAILED")
```

Is it too magic?